### ISSUE TITLE
IfcPatch ExtractElements: TypeError: 'type' object is not subscriptable

### ISSUE THREAD

**Issue by wlinna** [IfcOpenShell/IfcOpenShell#4546]
Hello, I'am trying to use IfcPatch to extract elements, and ExtractElements seems like it might do the job. However, it seems to be unable to perform. Here's how I tested it:

1. Install `ifcopenshell` and its dependency `lark` 
```
pip3 install lark ifcopenshell
```
2. Clone IfcOpenShell repository
4. Open IfcPatch directory in a terminal
5. Copy [this IFC file](https://raw.githubusercontent.com/andrewisen/bim-whale-ifc-samples/main/BasicHouse/IFC/BasicHouse.ifc) to the directory
6. Run `python3 -m ifcpatch -i BasicHouse.ifc -o BasicHouseSlabs.ifc -r ExtractElements -a "IfcSlab"`

Expected result: 
A file called BasicHouseSlabs.ifc

Actual result:
```
# Loading IFC file ...
# Patching ...
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "***/IfcOpenShell/src/ifcpatch/ifcpatch/__main__.py", line 37, in <module>
    output = ifcpatch.execute(args)
  File "***/IfcOpenShell/src/ifcpatch/ifcpatch/__init__.py", line 74, in execute
    recipes = getattr(__import__("ifcpatch.recipes.{}".format(args["recipe"])), "recipes")
  File "***/IfcOpenShell/src/ifcpatch/ifcpatch/recipes/ExtractElements.py", line 21, in <module>
    import ifcopenshell.util.selector
  File "***/.local/lib/python3.10/site-packages/ifcopenshell/util/selector.py", line 27, in <module>
    import ifcopenshell.util.placement
  File "***/.local/lib/python3.10/site-packages/ifcopenshell/util/placement.py", line 23, in <module>
    MatrixType = np.ndarray[np.ndarray[float]]
TypeError: 'type' object is not subscriptable

```

My environment:
- Linux pop-os 6.8.0 (64-bit)  (Pop!_OS is mostly Ubuntu-based) 
- Python 3.10.12

**Comment by aothms** [comment]
> MatrixType = np.ndarray[np.ndarray[float]]
> TypeError: 'type' object is not subscriptable

I didn't know that was possible. Must be a new feature of some recent numpy version. Can you try upgrading. A quick search online doesn't tell me much which version introduced this.

Honestly, we're getting a lot of compatibility issues for type annotations.

**Comment by wlinna** [comment]
Thanks

Thanks

`pip3 install --force numpy` got me version 1.26.4, seemed to help, although I still got this error:

```
  File "***/IfcOpenShell/src/ifcpatch/ifcpatch/recipes/ExtractElements.py", line 21, in <module>
    import ifcopenshell.util.selector
  File "***/.local/lib/python3.10/site-packages/ifcopenshell/util/selector.py", line 31, in <module>
    import ifcopenshell.util.shape
  File "***/.local/lib/python3.10/site-packages/ifcopenshell/util/shape.py", line 19, in <module>
    import shapely
ModuleNotFoundError: No module named 'shapely'
```

After `pip install shapely`, the process was able to finish without errors.

I haven't actively programmed in Python for a while, but I thought installing a package would usually install the dependencies as well, provided that they are listed in the dependencies in the first place. Is it possible that these dependencies have not been listed, or maybe I just made a mistake by directly installing ifcopenshell without creating venv or anything.

**Comment by wlinna** [comment]
However, getting the

**Comment by Moult** [comment]
IfcPatch is a collection of recipes, some with more onerous deps than others, so those are optional deps. I haven't gone through yet and specified them so let's keep this open until I fix that. Most users use ifcpatch through the blenderbim Add-on graphically where we bundle all deps.

That selector docs is outdated, see the official docs here:

https://docs.ifcopenshell.org/ifcopenshell-python/selector_syntax.html#filtering-elements

Tldr just remove the # prefix.

**Comment by wlinna** [comment]
Thanks again :D, with this new syntax I got what I wanted. Just in case someone ends up here by accident looking for this information, here's the command I used in the end

```
python3 -m ifcpatch -i BasicHouse.ifc -o BasicHouse2Pieces.ifc -r ExtractElements -a "1yETHMphv6LwABqR4Pbs5g,1yETHMphv6LwABqR0Pbs5g"   
```

> IfcPatch is a collection of recipes, some with more onerous deps than others, so those are optional deps

Okay, that makes sense

**Comment by wlinna** [comment]
> That selector docs is outdated, see the official docs here:
> https://docs.ifcopenshell.org/ifcopenshell-python/selector_syntax.html#filtering-elements

This page here links directly to the old documentation
https://docs.ifcopenshell.org/ifcpatch.html
